File: src/cli/roadmap/projects.ts
import type { Command } from "commander";
import { getClient } from "../../lib/client.ts";
import { printError, printPaginated, resolvePageSize } from "../../lib/output.ts";
import { resolveRoadmap } from "../../lib/resolvers.ts";

export function registerProjects(roadmap: Command): void {
  roadmap
    .command("projects")
    .description("List projects linked to a roadmap")
    .argument("<id>", "Roadmap ID, slug, or name")
    .option("--limit <n>", "Limit results", "50")
    .option("--cursor <token>", "Pagination cursor for next page")
    .action(async (id: string, opts: { limit: string; cursor?: string }) => {
      try {
        const client = getClient();
        const r = await resolveRoadmap(client, id);
        const projects = await r.projects({
          first: resolvePageSize(opts),
          after: opts.cursor,
        });
        const items = await Promise.all(
          projects.nodes.map(async (p) => {
            const lead = await p.lead;
            return {
              id: p.id,
              slugId: p.slugId,
              url: p.url,
              name: p.name,
              status: p.state,
              progress: p.progress,
              lead: lead ? lead.name : null,
              startDate: p.startDate,
              targetDate: p.targetDate,
            };
          }),
        );
        printPaginated(items, projects.pageInfo);
      } catch (err) {
        printError(err instanceof Error ? err.message : "Get projects failed");
      }
    });
}

File: internal/cli/roadmap/projects.go
package roadmap

import (
	"context"

	"github.com/spf13/cobra"

	"github.com/shhac/lin/internal/linear"
	"github.com/shhac/lin/internal/mappers"
	"github.com/shhac/lin/internal/output"
	"github.com/shhac/lin/internal/ptr"
	"github.com/shhac/lin/internal/resolvers"
)

func registerProjects(roadmap *cobra.Command) {
	var limit string
	var cursor string

	cmd := &cobra.Command{
		Use:   "projects <id>",
		Short: "List projects linked to a roadmap",
		Args:  cobra.ExactArgs(1),
		Run: func(_ *cobra.Command, args []string) {
			client := linear.GetClient()
			pageSize := output.ResolvePageSize(limit)

			resolved, err := resolvers.ResolveRoadmap(client, args[0])
			if err != nil {
				output.PrintError(err.Error())
			}

			after := output.ResolveCursor(cursor)

			resp, err := linear.RoadmapProjects(context.Background(), client, resolved.ID, pageSize, after)
			if err != nil {
				output.HandleGraphQLError(err)
			}

			items := make([]map[string]any, len(resp.Roadmap.Projects.Nodes))
			for i, p := range resp.Roadmap.Projects.Nodes {
				f := p.ProjectSummaryFields
				input := mappers.ProjectSummaryInput{
					ID:         f.Id,
					SlugId:     f.SlugId,
					URL:        f.Url,
					Name:       f.Name,
					State:      f.State,
					Progress:   f.Progress,
					StartDate:  ptr.Deref(f.StartDate),
					TargetDate: ptr.Deref(f.TargetDate),
				}
				if f.Lead != nil {
					input.LeadName = f.Lead.Name
				}
				items[i] = mappers.MapProjectSummary(input)
			}

			pi := resp.Roadmap.Projects.PageInfo
			output.PrintPaginated(items, &output.Pagination{
				HasMore:    pi.HasNextPage,
				NextCursor: ptr.Deref(pi.EndCursor),
			})
		},
	}

	cmd.Flags().StringVar(&limit, "limit", "50", "Limit results")
	cmd.Flags().StringVar(&cursor, "cursor", "", "Pagination cursor for next page")
	roadmap.AddCommand(cmd)
}
